geo/geo*fn: handle EMPTY cases in GEOMETRY/GEOGRAPHY

otan opened the pull request:
There are many cases of EMPTY that we need to handle for each method we
support. This includes `POINT EMPTY`, `LINESTRING EMPTY` ...
`GEOMETRYCOLLECTION EMPTY` (i.e. one for each shape). Furthermore,
there's an even more confusing GEOMETRYCOLLECTION with EMPTY shapes,
e.g. `GEOMETRYCOLLECTION ( LINESTRING EMPTY ) `

Both GEOS and PostGIS have various bugs in handling these, which have
been commented in code and imitated where appropriate.

Also bump twpayne/go-geom to pickup the fixed Empty() checks.

Resolves https://github.com/cockroachdb/cockroach/issues/49182, 

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/49247)
<!-- Reviewable:end -->

sumeerbhola reviewed (changes requested):
Looks good overall -- just some minor comments.

> Reviewed 19 of 19 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49247#-:-M8Gf7sA8vCzg1Mo81vL:b-acvuec)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan)

___
*[pkg/geo/geogfn/distance_test.go, line 282 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49247#-M8Gb7CQ9_NMhcMft9Kd:-M8Gb7CQ9_NMhcMft9Ke:b-lbfa83) ([raw file](https://github.com/cockroachdb/cockroach/blob/0de32eb9040c266a1e4417f15288e37757e6cf5c/pkg/geo/geogfn/distance_test.go#L282)):*
> ```Go
> 			{"GEOMETRYCOLLECTION EMPTY", "GEOMETRYCOLLECTION EMPTY"},
> 			{"GEOMETRYCOLLECTION EMPTY", "GEOMETRYCOLLECTION (POINT(1.0 1.0), LINESTRING EMPTY)"},
> 			{"POINT(1.0 1.0)", "GEOMETRYCOLLECTION (POINT(1.0 1.0), LINESTRING EMPTY)"}, // This case errors (in a bad way) in PostGIS.
> ```

what is the "bad way"?
___
*[pkg/geo/geogfn/dwithin.go, line 32 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49247#-M8GbVlU-XvNu3bGk9VD:-M8GbVlU-XvNu3bGk9VE:bag93qc) ([raw file](https://github.com/cockroachdb/cockroach/blob/0de32eb9040c266a1e4417f15288e37757e6cf5c/pkg/geo/geogfn/dwithin.go#L32)):*
> ```Go
> 	}
> 
> 	aRegions, err := a.AsS2(geo.EmptyBehaviorError)
> ```

Can you add a comment
// If any of a or b contain empty geometries, DWithin is false.
___
*[pkg/geo/geomfn/distance.go, line 24 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49247#-M8GcE8d4cP7o7NC5DG-:-M8GcE8d4cP7o7NC5DG0:b9swu0m) ([raw file](https://github.com/cockroachdb/cockroach/blob/0de32eb9040c266a1e4417f15288e37757e6cf5c/pkg/geo/geomfn/distance.go#L24)):*
> ```Go
> 
> // MinDistance returns the minimum distance between geometries A and B.
> // This can return a geo.EmptyGeometryError if either A or B is EMPTY.
> ```

nit:  This returns a ...
___
*[pkg/geo/geomfn/distance.go, line 52 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49247#-M8GcbL45TZoDfjDk1__:-M8GcbL45TZoDfjDk1_a:brt0rjk) ([raw file](https://github.com/cockroachdb/cockroach/blob/0de32eb9040c266a1e4417f15288e37757e6cf5c/pkg/geo/geomfn/distance.go#L52)):*
> ```Go
> 	if err != nil {
> 		// In case of ANY empty geometries return false (despite returning the
> 		// distance in MinDistance).
> ```

I don't think the parenthetical comment is necessary (and may be confusing), since this function never returns the distance.
___
*[pkg/geo/geomfn/distance.go, line 73 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49247#-M8GcuzRB4GKJjZVtmV8:-M8GcuzRB4GKJjZVtmV9:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/0de32eb9040c266a1e4417f15288e37757e6cf5c/pkg/geo/geomfn/distance.go#L73)):*
> ```Go
> 	if err != nil {
> 		// In case of ANY empty geometries return false (despite returning the
> 		// distance in MinDistance).
> ```

ditto
___
*[pkg/geo/geomfn/distance.go, line 124 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49247#-M8GdktRBk8zO1B4-0J0:-M8GdktRBk8zO1B4-0J1:b-psmbvu) ([raw file](https://github.com/cockroachdb/cockroach/blob/0de32eb9040c266a1e4417f15288e37757e6cf5c/pkg/geo/geomfn/distance.go#L124)):*
> ```Go
> 
> 	// This behaves differently than geogfn.Distance in that it will consider
> 	// non-empty elements.
> ```

I didn't understand this comment. Is it referring to the fact that for geography we never do `EmptyBehaviorOmit` so `flattenGeometry` will error out earlier, while with `EmptyBehaviorOmit` the code here will iterate over the non-empty elements?
If so, it would be better to make this comment stand-alone and not make a comparison with geography.
___
*[pkg/sql/sem/builtins/geo_builtins.go, line 806 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49247#-M8Gekti-T6-juTKlkIW:-M8Gekti-T6-juTKlkIX:bgql546) ([raw file](https://github.com/cockroachdb/cockroach/blob/0de32eb9040c266a1e4417f15288e37757e6cf5c/pkg/sql/sem/builtins/geo_builtins.go#L806)):*
> ```Go
> 				ret, err := geomfn.MinDistance(a.Geometry, b.Geometry)
> 				if err != nil {
> 					if geo.IsEmptyGeometryError(err) {
> ```

doesn't `geomfn.MinDistance` already do the appropriate error translation?
I'm confused by the error translations in this file. 


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49247)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @sumeerbhola)

___
*[pkg/geo/geogfn/distance_test.go, line 282 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49247#-M8Gb7CQ9_NMhcMft9Kd:-M8H1CKd0hRjKPasF-4H:b6ohn8n) ([raw file](https://github.com/cockroachdb/cockroach/blob/0de32eb9040c266a1e4417f15288e37757e6cf5c/pkg/geo/geogfn/distance_test.go#L282)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

what is the "bad way"?
</blockquote></details>

hmm, bad copy pasta.  removed comment.
___
*[pkg/geo/geogfn/dwithin.go, line 32 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49247#-M8GbVlU-XvNu3bGk9VD:-M8H3JUW6NVpihlXiNc9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0de32eb9040c266a1e4417f15288e37757e6cf5c/pkg/geo/geogfn/dwithin.go#L32)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Can you add a comment
// If any of a or b contain empty geometries, DWithin is false.
</blockquote></details>

Done.
___
*[pkg/geo/geomfn/distance.go, line 52 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49247#-M8GcbL45TZoDfjDk1__:-M8H1OayDwluQMfyaEPR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0de32eb9040c266a1e4417f15288e37757e6cf5c/pkg/geo/geomfn/distance.go#L52)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I don't think the parenthetical comment is necessary (and may be confusing), since this function never returns the distance.
</blockquote></details>

Done.
___
*[pkg/geo/geomfn/distance.go, line 73 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49247#-M8GcuzRB4GKJjZVtmV8:-M8H1UoK8ZhKXyyocq94:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0de32eb9040c266a1e4417f15288e37757e6cf5c/pkg/geo/geomfn/distance.go#L73)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Done.
___
*[pkg/geo/geomfn/distance.go, line 124 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49247#-M8GdktRBk8zO1B4-0J0:-M8H27LO3lYpA6Ul4RQM:bb220j6) ([raw file](https://github.com/cockroachdb/cockroach/blob/0de32eb9040c266a1e4417f15288e37757e6cf5c/pkg/geo/geomfn/distance.go#L124)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I didn't understand this comment. Is it referring to the fact that for geography we never do `EmptyBehaviorOmit` so `flattenGeometry` will error out earlier, while with `EmptyBehaviorOmit` the code here will iterate over the non-empty elements?
If so, it would be better to make this comment stand-alone and not make a comparison with geography.
</blockquote></details>

Fixed up the comment.
___
*[pkg/sql/sem/builtins/geo_builtins.go, line 806 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49247#-M8Gekti-T6-juTKlkIW:-M8H2J8x0T42FZKELSn5:b4qiai) ([raw file](https://github.com/cockroachdb/cockroach/blob/0de32eb9040c266a1e4417f15288e37757e6cf5c/pkg/sql/sem/builtins/geo_builtins.go#L806)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

doesn't `geomfn.MinDistance` already do the appropriate error translation?
I'm confused by the error translations in this file. 
</blockquote></details>

No -- because we want to return NULL here instead of 0. They have to be unhandled there to be handled here.


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 5 of 7 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49247#-:-M8HTCq-9zCn_TRoNpFS:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @otan and @sumeerbhola)

___
*[pkg/geo/geomfn/distance.go, line 103 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/49247#-M8HSn2FCXqg9zsZF1JB:-M8HSn2G1O6MOhCyvJON:be4vzcu) ([raw file](https://github.com/cockroachdb/cockroach/blob/23797655d2de0d94223baa50ab74c21f450e5f71/pkg/geo/geomfn/distance.go#L103)):*
> ```Go
> // distanceInternal calculates the distance between two geometries using
> // the DistanceCalculator operator.
> // If there are any EMPTY Geometry objects, they will be ignored. It will return an
> ```

isn't "they will be ignored" subject to the `emptyBehavior` parameter?


<!-- Sent from Reviewable.io -->

otan commented:
bors r=sumeerbhola

otan commented:
bors r-

otan committed:
geo/geo*fn: handle EMPTY cases in GEOMETRY/GEOGRAPHY

There are many cases of EMPTY that we need to handle for each method we
support. This includes `POINT EMPTY`, `LINESTRING EMPTY` ...
`GEOMETRYCOLLECTION EMPTY` (i.e. one for each shape). Furthermore,
there's an even more confusing GEOMETRYCOLLECTION with EMPTY shapes,
e.g. `GEOMETRYCOLLECTION ( LINESTRING EMPTY ) `

Both GEOS and PostGIS have various bugs in handling these, which have
been commented in code and imitated where appropriate.

Also bump twpayne/go-geom to pickup the fixed Empty() checks.

Release note: None

otan commented:
bors r=sumeerbhola

sumeerbhola reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49247)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @otan and @sumeerbhola)

___
*[pkg/geo/geomfn/distance.go, line 103 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/49247#-M8HSn2FCXqg9zsZF1JB:-M8LEqMNBObyFEV2Q3oi:b-1tjqsv) ([raw file](https://github.com/cockroachdb/cockroach/blob/23797655d2de0d94223baa50ab74c21f450e5f71/pkg/geo/geomfn/distance.go#L103)):*
> ```
>  If there are any EMPTY Geometry objects, they will be ignored
> ```

If `distanceInternal(..., geo.EmptyBehaviorError)` is called with `a` containing a collection of geometry objects only one of which is empty, will this function return an error?


<!-- Sent from Reviewable.io -->
